packages/init/node-async_hooks.js

```diff
@@ -0,0 +1,226 @@
+// @ts-check
+
+import { createHook, AsyncResource } from 'async_hooks';
+
+/// <reference path="./node-async_hooks-symbols.d.ts" />
+
+const asyncHooksWellKnownNameFromDescription = {
+  async_id_symbol: 'nodeAsyncHooksAsyncId',
+  trigger_async_id_symbol: 'nodeAsyncHooksTriggerAsyncId',
+  destroyed: 'nodeAsyncHooksDestroyed',
+};
+
+const promiseAsyncHookFallbackStates = new WeakMap();
+
+const setAsyncSymbol = (description, symbol) => {
+  const wellKnownName = asyncHooksWellKnownNameFromDescription[description];
+  if (!wellKnownName) {
+    throw new Error('Unknown symbol');
+  } else if (!Symbol[wellKnownName]) {
+    Symbol[wellKnownName] = symbol;
+    return true;
+  } else if (Symbol[wellKnownName] !== symbol) {
+    // console.warn(
+    //   `Found duplicate ${description}:`,
+    //   symbol,
+    //   Symbol[wellKnownName],
+    // );
+    return false;
+  } else {
+    return true;
+  }
+};
+
+// We can get the `async_id_symbol` and `trigger_async_id_symbol` through a
+// simple instantiation of async_hook.AsyncResource, which causes little side
+// effects. These are the 2 symbols that may be late bound, aka after the promise
+// is returned to the program and would normally be frozen.
+const findAsyncSymbolsFromAsyncResource = () => {
+  let found = 0;
+  Object.getOwnPropertySymbols(new AsyncResource('Bootstrap')).forEach(sym => {
+    const { description } = sym;
+    if (description in asyncHooksWellKnownNameFromDescription) {
+      if (setAsyncSymbol(description, sym)) {
+        found += 1;
+      }
+    }
+  });
+  return found;
+};
+
+// To get the `destroyed` symbol installed on promises by async_hooks,
+// the only option is to create and enable an AsyncHook.
+// Different versions of Node handle this in various ways.
+const findAsyncSymbolsFromPromiseCreateHook = () => {
+  const bootstrapData = [];
+
+  {
+    const bootstrapHook = createHook({
+      init(asyncId, type, triggerAsyncId, resource) {
+        if (type !== 'PROMISE') return;
+        // console.log('Bootstrap', asyncId, triggerAsyncId, resource);
+        bootstrapData.push({ asyncId, triggerAsyncId, resource });
+      },
+      destroy(_asyncId) {
+        // Needs to be present to trigger the addition of the destroyed symbol
+      },
+    });
+
+    bootstrapHook.enable();
+    // Use a never resolving promise to avoid triggering settlement hooks
+    const trigger = new Promise(() => {});
+    bootstrapHook.disable();
+
+    // In some versions of Node, async_hooks don't give access to the resource
+    // itself, but to a "wrapper" which is basically hooks metadata for the promise
+    const promisesData = bootstrapData.filter(
+      ({ resource }) => Promise.resolve(resource) === resource,
+    );
+    bootstrapData.length = 0;
+    const { length } = promisesData;
+    if (length > 1) {
+      // console.warn('Found multiple potential candidates');
```

Comment:
Are you aware of a situation when this would happen? AFAIR it shouldn't be possible.

Reply:
I did see it happen at one point when I was exploring, and then couldn't reproduce. not sure why exactly, but I kept the check around.

Reply:
It would mean the init hook was called twice. I'd consider that a bug.

Reply:
I think it created multiple promise instances for some reason.